### ISSUE TITLE
mountOnEnter / preEnter issue

### ISSUE THREAD

**Issue by guitarhero17** [szhsin/react-transition-state#411]
Hello,

I've stumbled across your library and immediately found it intriguing and a very good replacement for `react-transition-group`. Unfortunately, I am currently experiencing an issue related to the enter transition - it does not always fire correctly.

I use a `Next.js` app. I managed to reproduce the false behavior in this [CodeSandbox](https://codesandbox.io/p/sandbox/sharp-sound-jzfh5j). 

As you can see, the `isLoading` prop of `Child` is toggled by a parent component. Whenever it changes, I fire the `useEffect` hook to toggle the transition. However, most of the time the enter transition (from opacity 0 to 1) is not performed at all but the opacity is rather changed in an instant. 

If `mountOnEnter` is false as per default and the element is mounted eagerly, the transition is performed correctly. I just have to make sure that `opacity: 0;` is there in the `hey` class.

Could you tell me what I'm doing wrong here?

Maybe the issue is related to `Next.js` as I tried the exact same thing in a `React` sandbox and couldn't reproduce it. 

I'd appreciate your assistance, thank you!

**Comment by alekseykarpenko** [comment]
+1 same issue with Next.js / React 18

**Comment by szhsin** [comment]
Hi @guitarhero17 , thanks for raising this issue. Upon investigation, I found it was related to where the toggle function was called. Still, I didn't fully understand the culprit but managed to find a solution that worked for most of time.

> TL;DR - When creating transition as a result of side-effect from another state update, we need to call toggle function from `useLayoutEffect` instead of `useEffect`.

To use CSS transition, we enable the pre-enter config to add a `preEnter` state immediate prior to the `entering` state. Ideally the `entering` state update should happen at the next tick and we use a zero ms timer to achieve it:
```js
// A simplified implementation

if (state === 'preEnter') setTimeout(() => setState('entering'), 0)
```
Although this will guarantee two render passes - first with 'preEnter' and then 'entering' states, it doesn't always result in two individual DOM updates. I reckon that `useEffect` has optimisation which batches consecutive DOM update into one go if those updates are very close in terms of timing. When there is only one DOM update, we cannot see a transition.

In contrast, `useLayoutEffect`, according to [its documentation](https://react.dev/reference/react/useLayoutEffect#caveats), makes sure a state update scheduled from it block the browser from repainting the screen, until DOM updates are completed. This causes React to flush the `preEnter` state into DOM before `entering` state update is scheduled. We will get two DOM update in the end.

I did some tests by replacing `useEffect` with `useLayoutEffect` with both React 17 and 18, and the results were promising and quite consistent - [CodeSandbox](https://codesandbox.io/p/sandbox/react-transition-state-411-kj2qkt)

**Comment by guitarhero17** [comment]
Hey @szhsin , thank you for your thorough response. `useLayoutEffect` does look promising, however I'm still experiencing the error on Firefox. The Chromium browsers and Safari seem to handle `preEnter` correctly but it looks like Firefox 114.0.2 (arm64 on a Macbook Air M1) does not consistently provide two render passes. The enter transition is shown only sometimes. To verify, open your [CodeSandbox](https://codesandbox.io/p/sandbox/react-transition-state-411-kj2qkt) example in the latest stable Firefox.

Thanks for your insights!

**Comment by szhsin** [comment]
Thanks @guitarhero17 . I can also reproduce the issue on Firefox. Will be looking into it further and try to figure out another solution.

**Comment by szhsin** [comment]
Hey folks, I managed to use a trick that forces browser to repaint after `preEnter`, and the issue has been fixed for me on all three browsers. If you still experience the issue please let me know.
- [CodeSandbox](https://codesandbox.io/p/sandbox/react-transition-state-411-kj2qkt)
- [2.1.1-alpha.1](https://www.npmjs.com/package/react-transition-state/v/2.1.1-alpha.1) `npm i react-transition-state@2.1.1-alpha.1`
- [Fixing branch](https://github.com/szhsin/react-transition-state/commit/1896107023ff85cfe1cde4428ba2957b35addcdc)

**Comment by alekseykarpenko** [comment]
@szhsin Thanks for jumping in so quickly! Your lib is amazing.

**Comment by szhsin** [comment]
Fixed by [v2.1.1](https://github.com/szhsin/react-transition-state/releases/tag/v2.1.1)

**Comment by guitarhero17** [comment]
Thank you for the fix @szhsin ! I don't see the problem anymore, it works correctly on all browsers I've tested 🎉  🙌🏻

**Comment by guitarhero17** [comment]
Hey @szhsin , just a quick follow-up question - do we still need to use `useLayoutEffect` or the recent changes will trigger a repaint even if we `trigger()` inside of  `useEffect` ?

**Comment by szhsin** [comment]
Hi @alekseykarpenko , as I tested on different browsers, the latest fix worked great on Chrome/Firebox by toggling states inside `useEffect`, but the issue still occurred on Safari unless states were toggled inside `useLayoutEffect`.

Combining the latest fix and  `useLayoutEffect`, we can make sure it works on the three browsers.
